web/src/peer_data.ts

```diff
@@ -1,22 +1,109 @@
+import assert from "minimalistic-assert";
+import {z} from "zod";
+
 import * as blueslip from "./blueslip.ts";
+import * as channel from "./channel.ts";
 import {LazySet} from "./lazy_set.ts";
+import {page_params} from "./page_params.ts";
 import type {User} from "./people.ts";
 import * as people from "./people.ts";
 import * as sub_store from "./sub_store.ts";
+import * as util from "./util.ts";
 
 // This maps a stream_id to a LazySet of user_ids who are subscribed.
 const stream_subscribers = new Map<number, LazySet>();
 const fetched_stream_ids = new Set<number>();
+export function has_full_subscriber_data(stream_id: number): boolean {
+    return fetched_stream_ids.has(stream_id);
+}
+const pending_subscriber_requests = new Map<
+    number,
+    {
+        subscribers_promise: Promise<LazySet | null>;
+        pending_peer_events: {
+            type: "peer_add" | "peer_remove";
+            user_ids: number[];
+        }[];
+    }
+>();
 
 export function clear_for_testing(): void {
     stream_subscribers.clear();
+    fetched_stream_ids.clear();
+}
+
+const fetch_stream_subscribers_response_schema = z.object({
+    subscribers: z.array(z.number()),
+});
+
+// If `retry_on_failure` is `true`, this function will always
+// resolve to a LazySet but could hang indefinitely trying to
+// fetch data. `num_attempts` should only be set in the recursive
+// call.
+// If `retry_on_failure` is `false`, this function can resolve
+// to `null` if there's an error during the fetch.
+export async function maybe_fetch_stream_subscribers(
+    stream_id: number,
+    retry_on_failure: boolean,
+    num_attempts = 1,
+): Promise<LazySet | null> {
+    if (pending_subscriber_requests.has(stream_id)) {
+        return pending_subscriber_requests.get(stream_id)!.subscribers_promise;
+    }
+    const subscribers_promise = (async () => {
+        let subscribers: number[];
+        let xhr: JQuery.jqXHR<unknown> | undefined;
+        try {
+            // TODO: Once we await `channel.get` we get a response
+            // of type unknown. But we also need jqXHR to caculate
+            // `get_retry_backoff_seconds`. How do we get the `xhr`
+            // for `get` errors but also properly return the right
+            // value from this function?
+            xhr = await channel.get({
+                url: `/json/streams/${stream_id}/members`,
+            });
+            subscribers = fetch_stream_subscribers_response_schema.parse(xhr).subscribers;
```

Comment:
There isn’t a rule that you have to `await` an expression as soon as it appears.

```js
xhr = channel.get({url: `/json/streams/${stream_id}/members`});
subscribers = fetch_stream_subscribers_response_schema.parse(await xhr).subscribers;
```